pkg/cri/io/metrics.go

```diff
@@ -0,0 +1,42 @@
+/*
+   Copyright The containerd Authors.
+
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+       http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package io
+
+import "github.com/docker/go-metrics"
+
+var (
+	inputEntries  metrics.Counter
+	outputEntries metrics.Counter
+	inputBytes    metrics.Counter
+	outputBytes   metrics.Counter
+	splitEntries  metrics.Counter
+)
+
+func init() {
+	// These CRI metrics record input and output logging volume.
+	ns := metrics.NewNamespace("containerd", "cri", nil)
+
+	inputEntries = ns.NewCounter("input_entries", "Number of log entries received")
+	outputEntries = ns.NewCounter("output_entries", "Number of log entries successfully written to disks")
+	inputBytes = ns.NewCounter("input_bytes", "Size of logs received")
+	outputBytes = ns.NewCounter("output_bytes", "Size of logs successfully written to disks")
+	splitEntries = ns.NewCounter("split_entries", "Number of extra log entries created by splitting the "+
```

Comment:
Regarding split_entries, would we expect users to check and reduce the # of split entries? If so, we may need to add some logging to help users to find problematic containers.

Reply:
Actually not. Logs getting split by Containerd is completely fine for us (and for user workloads). However, because of this behavior of Containerd, the number of output entries will usually be greater than the number of input entries. In this case, we cannot tell if the extra output entries are caused by Containerd duplicate logs somewhere or by Containerd split log entries. Adding split_entries helps us distinguish these two cases. Ideally input_entries + split_entries = output_entries. I added more design details in the PR description which covers this case.

Reply:
Thanks for the detailed explanation!